editor/editor_node.cpp

```diff
@@ -606,8 +606,10 @@ void EditorNode::_notification(int p_what) {
 				_editor_select(EDITOR_3D);
 			}
 
-			// Save the project after opening to mark it as last modified.
-			ProjectSettings::get_singleton()->save();
+			// Save the project after opening to mark it as last modified, except in headless mode.
+			if (!OS::get_singleton()->is_no_window_mode_enabled()) {
```

Comment:
I don't think this will be sufficient when running with the headless display server without the (now mostly redundant) `--no-window` flag.

Reply:
Case in point, we removed the `--no-window` flag in #51216 :)

Checking `OS::get_singleton()->can_draw()` might be better. (CC @Faless)

Reply:
> _Edit:_ That's for `3.x`, in `master` it would likely be `DisplayServer::get_singleton()->window_can_draw()`.

Pushed an amend with that.

Reply:
I tested on `3.x` and there the no window mode doesn't set `can_draw()` to false, so we need to check both:
```
OS::get_singleton()->can_draw() && !OS::get_singleton()->is_no_window_mode_enabled()
```
(to exclude both server and --no-window desktop builds)